### PR TITLE
chore: dispose stale handles to prevent oom, 1000 of a kind max

### PR DESCRIPTION
**Pull request by pavelfeldman**
https://github.com/microsoft/playwright/issues/6319

### CODE REVIEW

File: packages/playwright-core/src/server/dispatchers/dispatcher.ts [hunk 1]
```diff
@@ -55,18 +60,19 @@ export class Dispatcher<Type extends { guid: string }, ChannelType, ParentScopeT
     this._parent = parent instanceof DispatcherConnection ? undefined : parent;
 
     const guid = object.guid;
+    this._guid = guid;
+    this._type = type;
+    this._object = object;
+
+    (object as any)[dispatcherSymbol] = this;
+
     assert(!this._connection._dispatchers.has(guid));
```
Comment (dgozman) [inline packages/playwright-core/src/server/dispatchers/dispatcher.ts:69]: nit: move assert to `registerDispatcher`?

File: packages/playwright-core/src/server/dispatchers/dispatcher.ts [hunk 2]
```diff
@@ -224,6 +233,31 @@ export class DispatcherConnection {
...
+    }
+    list.add(dispatcher._guid);
+    if (list.size > maxDispatchers)
+      this._disposeStaleDispatchers(list);
+  }
+
+  private _disposeStaleDispatchers(dispatchers: Set<string>) {
+    const dispatchersArray = [...dispatchers];
+    const collect = dispatchersArray.splice(0, maxDispatchers / 10);
+    dispatchers = new Set(dispatchersArray.slice(maxDispatchers / 10));
```
Comment (dgozman) [inline packages/playwright-core/src/server/dispatchers/dispatcher.ts:252]: What does this line do? Did you mean to mutate `dispatchers` set in-place? Probably not!

File: tests/stress/heap.spec.ts [hunk 3]
```diff
@@ -18,6 +18,7 @@ import { contextTest as test, expect } from '../config/browserTest';
 import { queryObjectCount } from '../config/queryObjects';
 
 test.describe.configure({ mode: 'serial' });
+test.skip(({ browserName }) => browserName !== 'chromium');
 
 for (let i = 0; i < 3; ++i) {
   test(`test #${i} to request page and context`, async ({ page, context }) => {
```
Comment (dgozman) [inline tests/stress/heap.spec.ts:24]: How is this different from line 29?

File: tests/stress/heap.spec.ts [hunk 4]
```diff
@@ -65,7 +66,7 @@ test('should not leak dispatchers after closing page', async ({ context, server
   expect(await queryObjectCount(require('../../packages/playwright-core/lib/server/page').Page)).toBe(COUNT);
   expect(await queryObjectCount(require('../../packages/playwright-core/lib/server/dispatchers/networkDispatchers').RequestDispatcher)).toBe(COUNT);
   expect(await queryObjectCount(require('../../packages/playwright-core/lib/server/dispatchers/networkDispatchers').ResponseDispatcher)).toBe(COUNT);
-  expect(await queryObjectCount(require('../../packages/playwright-core/lib/server/dispatchers/consoleMessageDispatcher').ConsoleMessageDispatcher)).toBe(COUNT);
+  expect(await queryObjectCount(require('../../packages/playwright-core/lib/server/dispatchers/consoleMessageDispatcher').ConsoleMessageDispatcher)).toBe(0);
```
Comment (dgozman) [inline tests/stress/heap.spec.ts:69]: There is no such thing as `ConsoleMessageDispatcher` anymore, so this line should throw all the time 😄

File: tests/stress/heap.spec.ts [hunk 5]
```diff
@@ -18,6 +18,7 @@ import { contextTest as test, expect } from '../config/browserTest';
 import { queryObjectCount } from '../config/queryObjects';
 
 test.describe.configure({ mode: 'serial' });
+test.skip(({ browserName }) => browserName !== 'chromium');
```
Comment (dgozman) [inline tests/stress/heap.spec.ts:21]: I wonder why is this chrome-specific?

**Review by dgozman — APPROVED** [review]

File: tests/stress/heap.spec.ts (hunk 3 above)
Comment (pavelfeldman) [reply tests/stress/heap.spec.ts:24]: :shrug:

**Comment by Gby56** [comment]
Hello @pavelfeldman 
I'm sorry for digging back up this PR, but I just wanted to ask if this would also be required in the python version of Playwright ?

I work with it, and we observe a lot of OOM due to long living contexts.

I can definitely work on that and make the PR if guided 😃 
Thank you for all your work !

**Comment by mxschmitt** [comment]
This was implemented in Python as well - if you experience any issues there, I recommend filing anew separate issue with reproduction steps. Thanks!

**Pull request merged**